modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/WeightedProspectSample.scala

```diff
@@ -0,0 +1,51 @@
+package org.tessellation.sdk.infrastructure.snapshot
+
+import cats.MonadThrow
+import cats.data.Chain
+import cats.effect.std.Random
+import cats.syntax.applicative._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.schema.trust.TrustValueRefined
+import org.tessellation.sdk.infrastructure.snapshot.WeightedSampling.{
+  ProbabilityRefinement,
+  WeightRefined,
+  WeightRefinement,
+  sample => weightedSample
+}
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.refineV
+import eu.timepit.refined.types.numeric.PosInt
+
+object WeightedProspectSample {
+
+  def sample[F[_]: MonadThrow: Random, A](candidates: Map[A, TrustValueRefined], sampleSize: PosInt): F[List[A]] = {
+    val filtered = candidates.view
+      .mapValues(score => refineV[WeightRefinement](score.value))
+      .collect {
+        case (key, Right(refinedScore)) =>
+          key -> refinedScore
+      }
+      .toMap
+
+    if (filtered.size <= sampleSize)
+      filtered.keySet.toList.pure
+    else {
+      type Result = List[A]
+      type Agg = (Chain[A], Map[A, WeightRefined])
+
+      val getP = Random[F].nextDouble.map(refineV[ProbabilityRefinement].unsafeFrom(_))
```

Comment:
using this, you would never get `1.0` value

Reply:
That is fine. Both `Random[F].nextDouble` and `ProbabilityRefinement` are 0 inclusive to 1 exclusive.

Reply:
Ok :+1: